### ISSUE TITLE
"torch.sum()" over a ByteTensor gives incorrect result when the "dim" is specified

### ISSUE THREAD

**Issue by sumitbinnani** [pytorch/pytorch#5863]
```
a = ByteTensor(40,500) #inititialized to some random values

print(torch.sum(a))
>>> 8186

print(torch.sum(torch.sum(a, dim=1)))
>>> 5626
```

The answer is incorrect as the `torch.sum(a, dim=1)` returns a ByteTensor resulting in a overflow.

**Comment by vishwakftw** [comment]
The behavior of `ByteTensor` has changed on `master`. Overflows are taken care of, and hence the value is restricted to the range [0, 255].

**Comment by vishwakftw** [comment]
http://pytorch.org/docs/master/tensors.html

The documentation specifies that this is 8-bit.

**Comment by sumitbinnani** [comment]
I am asking for a `warning` as previously `torch.sum()` was returning `LongTensor`. This change might affect a few existing codes as now the return type of the `torch.sum()` would be a `ByteTensor`.

**Comment by vishwakftw** [comment]
Please refer to this https://github.com/pytorch/pytorch/issues/5673.

**Comment by soumith** [comment]
cc: @gchanan can you take a decision / action on this.

**Comment by gchanan** [comment]
It's on my list -- I'll post a plan here once ready.

**Comment by mattiaspaul** [comment]
this might not be completely related, but torch.sum also gives unexpected results on large float tensors the first example works fine:
```
T = torch.randn(4, 100, 320)
M = torch.mean(T,0)*4.0
S = torch.sum(T,0)
print(torch.norm(M-S))
 0
[torch.FloatTensor of size ()]
```
But increasing the tensor size leads to a wrong result:
 ```
T = torch.randn(4, 200, 320)
M = torch.mean(T,0)*4.0
S = torch.sum(T,0)
print(torch.norm(M-S))
  507.3207
[torch.FloatTensor of size ()]
```
I have seen this problem when compiling the newest master version from source on both Linux and Mac.

**Comment by nikhilweee** [comment]
A warning would be greatly appreciated. I've spent hours debugging my code to finally figure out the issue. It's a common practise to use something like `correct_predictions = torch.eq(predictions, targets).sum()` and to find out that `correct_predictions` is always going to be less than 255 isn't intuitive.

**Comment by fmassa** [comment]
This has been fixed in 0.4. Now reductions on byte tensors return the result as a long tensor (except if the explicit `dtype` is passed)
